File: src/reminder.ts
import '@jxa/global-type';
import { run } from '@jxa/run';
import { SingletonService, Logger, JXAObject } from './shared.js';
import { promisify } from 'util';
import { exec } from 'child_process';

const execAsync = promisify(exec);

export type Reminder = {
  id: string;
  name: string;
  body: string;
  completed: boolean;
  completedDate: Date | null;
  dueDate: Date | null;
  alldayDueDate: Date | null;
  remindMeDate: Date | null;
  priority: number;
  flagged: boolean;
};

export type ReminderFilter = {
  completed?: boolean;
  flagged?: boolean;
  priority?: number;
  dueAfter?: Date;
  dueBefore?: Date;
};

// Type for JXA Reminder object
interface JXAReminder {
  // Read methods
  id(): string;
  name(): string;
  body(): string;
  completed(): boolean;
  completionDate(): Date | null;
  dueDate(): Date | null;
  alldayDueDate(): Date | null;
  remindMeDate(): Date | null;
  priority(): number;
  flagged(): boolean;

  // Write properties
  setName(value: string): void;
  setBody(value: string): void;
  setCompleted(value: boolean): void;
  setPriority(value: number): void;
  setFlagged(value: boolean): void;
  setDueDate(value: Date | null): void;
  setRemindMeDate(value: Date | null): void;
}

interface JXAList {
  name(): string;
  reminders(): JXAReminder[];
}

export class RemindersService extends SingletonService<Reminder> {
  private logger: Logger;

  constructor() {
    super();
    this.logger = Logger.getInstance();
  }

  protected convertFromJXA(jxaObject: JXAObject): Reminder {
    const reminder = jxaObject as unknown as JXAReminder;
    return {
      id: reminder.id(),
      name: reminder.name(),
      body: reminder.body(),
      completed: reminder.completed(),
      completedDate: reminder.completionDate(),
      dueDate: reminder.dueDate(),
      alldayDueDate: reminder.alldayDueDate(),
      remindMeDate: reminder.remindMeDate(),
      priority: reminder.priority(),
      flagged: reminder.flagged(),
    };
  }

  // Direct osascript implementation
  async createReminderDirect(name: string, body: string): Promise<Reminder> {
    try {
      this.logger.log('Creating reminder (direct)', { name, body });
      
      const script = `
        const app = Application('Reminders');
        const defaultList = app.defaultList();
        if (!defaultList) {
          throw new Error('No default list found');
        }
        
        const newReminder = defaultList.make({
          new: 'reminder',
          withProperties: {
            name: ${JSON.stringify(name)},
            body: ${JSON.stringify(body)},
            completed: false,
            priority: 0,
            flagged: false
          }
        });
        
        JSON.stringify({
          id: newReminder.id(),
          name: newReminder.name(),
          body: newReminder.body(),
          completed: newReminder.completed(),
          completedDate: newReminder.completionDate(),
          dueDate: newReminder.dueDate(),
          alldayDueDate: newReminder.alldayDueDate(),
          remindMeDate: newReminder.remindMeDate(),
          priority: newReminder.priority(),
          flagged: newReminder.flagged()
        });
      `;

      // Use spawn to pipe the script via stdin
      const { spawn } = require('child_process');
      //const child = spawn('osascript', ['-l', 'JavaScript', '-']);
      // launch echo in a shell to test
      const child = spawn('echo', ['hello']);
      this.logger.log(' echo test completed');
      return null;
      
      let stdout = '';
      let stderr = '';
      
      child.stdout.on('data', (data: Buffer) => {
        stdout += data.toString();
      });
      
      child.stderr.on('data', (data: Buffer) => {
        stderr += data.toString();
      });
      
      // Wrap in a promise to handle async completion
      const result = await new Promise<Reminder>((resolve, reject) => {
        child.on('close', (code: number) => {
          if (code !== 0) {
            reject(new Error(`osascript failed with code ${code}: ${stderr}`));
          } else {
            try {
              resolve(JSON.parse(stdout) as Reminder);
            } catch (e) {
              reject(new Error(`Failed to parse osascript output: ${e}`));
            }
          }
        });
        
        // Write script to stdin and close it
        child.stdin.write(script);
        child.stdin.end();
      });
      
      this.logger.log('Created reminder (direct)', result);
      return result;
    } catch (error) {
      await this.logger.logError(error, 'Failed to create reminder (direct)');
      throw error;
    }
  }

  async createReminder(name: string, body: string): Promise<Reminder> {
    try {
      this.logger.log('Creating reminder', { name, body });
      const result = await run(
        (name, body) => {
          console.log('Getting default list');
          const app = Application('Reminders');
          const defaultList = app.defaultList();
          if (!defaultList) {
            throw new Error('No default list found');
          }

          // Create with all properties at once to minimize property access
          console.log('Calling defaultList.make');
          const newReminder = defaultList.make({
            new: 'reminder',
            withProperties: {
              name,
              body,
              completed: false,
              priority: 0,
              flagged: false
            },
          }) as JXAReminder;
          console.log('Finished calling defaultList.make');

          // Return properties directly
          return {
            id: newReminder.id(),
            name: newReminder.name(),
            body: newReminder.body(),
            completed: newReminder.completed(),
            completedDate: newReminder.completionDate(),
            dueDate: newReminder.dueDate(),
            alldayDueDate: newReminder.alldayDueDate(),
            remindMeDate: newReminder.remindMeDate(),
            priority: newReminder.priority(),
            flagged: newReminder.flagged(),
          };
        },
        name,
        body,
      ) as Reminder;
      
      this.logger.log('Created reminder', result);
      return result;
    } catch (error) {
      await this.logger.logError(error, 'Failed to create reminder');
      throw error;
    }
  }

  async listReminders(filter?: ReminderFilter): Promise<Reminder[]> {
    try {
      this.logger.log('Listing reminders', { filter });
      
      const result = await run((filterStr) => {
        try {
          const filter = filterStr ? JSON.parse(filterStr) : {};
          const app = Application('Reminders');
          const lists = app.lists();

          // Get all reminders and their properties in bulk
          const allReminders = lists.flatMap((list: JXAList) => {
            const reminders = list.reminders();
            return reminders.map((r: JXAReminder) => ({
              id: r.id(),
              name: r.name(),
              body: r.body(),
              completed: r.completed(),
              completedDate: r.completionDate(),
              dueDate: r.dueDate(),
              alldayDueDate: r.alldayDueDate(),
              remindMeDate: r.remindMeDate(),
              priority: r.priority(),
              flagged: r.flagged(),
            }));
          });

          // Filter the reminders (all properties already retrieved)
          return allReminders.filter((reminder: Reminder) => {
            if (filter.completed !== undefined && reminder.completed !== filter.completed) {
              return false;
            }
            if (filter.flagged !== undefined && reminder.flagged !== filter.flagged) {
              return false;
            }
            if (filter.priority !== undefined && reminder.priority !== filter.priority) {
              return false;
            }
            if (filter.dueAfter && reminder.dueDate) {
              if (new Date(reminder.dueDate) < new Date(filter.dueAfter)) {
                return false;
              }
            }
            if (filter.dueBefore && reminder.dueDate) {
              if (new Date(reminder.dueDate) > new Date(filter.dueBefore)) {
                return false;
              }
            }
            return true;
          });
        } catch (e) {
          const error = e as { message?: string };
          throw new Error(`Failed to list reminders: ${error.message}`);
        }
      }, filter ? JSON.stringify(filter) : undefined) as Reminder[];

      this.logger.log('Listed reminders', { count: result.length });
      return result;
    } catch (error) {
      await this.logger.logError(error, 'Failed to list reminders');
      throw error;
    }
  }

  async updateReminder(id: string, updates: Partial<Omit<Reminder, 'id'>>): Promise<Reminder> {
    const result = await run(
      (id, updatesStr) => {
        const updates = JSON.parse(updatesStr);
        const app = Application('Reminders');
        const lists = app.lists();
        
        // Find reminder
        const reminder = lists
          .flatMap((list: JXAList) => list.reminders())
          .find((r: JXAReminder) => r.id() === id);

        if (!reminder) {
          throw new Error(`Reminder with id ${id} not found`);
        }

        // Apply updates directly
        if (updates.name !== undefined) reminder.setName(updates.name);
        if (updates.body !== undefined) reminder.setBody(updates.body);
        if (updates.completed !== undefined) reminder.setCompleted(updates.completed);
        if (updates.priority !== undefined) reminder.setPriority(updates.priority);
        if (updates.flagged !== undefined) reminder.setFlagged(updates.flagged);
        if (updates.dueDate !== undefined) reminder.setDueDate(updates.dueDate);
        if (updates.remindMeDate !== undefined) reminder.setRemindMeDate(updates.remindMeDate);

        // Return current properties
        return {
          id: reminder.id(),
          name: reminder.name(),
          body: reminder.body(),
          completed: reminder.completed(),
          completedDate: reminder.completionDate(),
          dueDate: reminder.dueDate(),
          alldayDueDate: reminder.alldayDueDate(),
          remindMeDate: reminder.remindMeDate(),
          priority: reminder.priority(),
          flagged: reminder.flagged(),
        };
      },
      id,
      JSON.stringify(updates),
    ) as Reminder;

    return result;
  }

  async completeReminder(id: string): Promise<Reminder> {
    return this.updateReminder(id, { completed: true });
  }

  async flagReminder(id: string, flagged: boolean): Promise<Reminder> {
    return this.updateReminder(id, { flagged });
  }

  async setPriority(id: string, priority: number): Promise<Reminder> {
    return this.updateReminder(id, { priority });
  }
}

File: src/reminder.py
#!/usr/bin/env python3

from Foundation import NSObject
from ScriptingBridge import SBApplication
import json
import sys
import logging
from subprocess import Popen, DEVNULL, PIPE
import tempfile

logging.basicConfig(level=logging.DEBUG, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

class RemindersService:
    def __init__(self):
        self.app = SBApplication.applicationWithBundleIdentifier_("com.apple.reminders")
        if not self.app:
            raise RuntimeError("Could not connect to Reminders.app")

    def create_reminder(self, name: str, body: str) -> dict:
        logger.debug("Creating reminder: %s", name)
        try:
            default_list = self.app.defaultList()
            if not default_list:
                raise RuntimeError("No default list found")
            
            logger.debug("Getting default list: %s", default_list.name())
            reminders = default_list.reminders()
            
            # Create reminder directly through Scripting Bridge
            reminder = self.app.classForScriptingClass_("reminder").alloc().init()
            
            # Get the current count of reminders
            initial_count = len(reminders)
            
            # Add the reminder to the list first
            reminders.addObject_(reminder)
            
            # Then set its properties
            reminder.setValue_forKey_(name, "name")
            reminder.setValue_forKey_(body, "body")
            
            # Verify the reminder was added
            new_count = len(reminders)
            if new_count <= initial_count:
                raise RuntimeError("Failed to add reminder to list")
            
            # Verify we can read back the properties
            if reminder.valueForKey_("name") != name:
                raise RuntimeError("Failed to set reminder name")
            
            return {
                "id": reminder.valueForKey_("id"),
                "name": reminder.valueForKey_("name"),
                "body": reminder.valueForKey_("body"),
                "completed": reminder.valueForKey_("completed"),
                "completionDate": reminder.valueForKey_("completionDate"),
                "dueDate": reminder.valueForKey_("dueDate"),
                "alldayDueDate": reminder.valueForKey_("alldayDueDate"),
                "remindMeDate": reminder.valueForKey_("remindMeDate"),
                "priority": reminder.valueForKey_("priority"),
                "flagged": reminder.valueForKey_("flagged")
            }
        except Exception as e:
            logger.error("Failed to create reminder: %s", e)
            raise

    def list_reminders(self, filter_params: dict = None) -> list:
        logger.debug("Listing reminders with filter: %s", filter_params)
        try:
            all_reminders = []
            for lst in self.app.lists():
                for reminder in lst.reminders():
                    reminder_dict = {
                        "id": reminder.id(),
                        "name": reminder.name(),
                        "body": reminder.body(),
                        "completed": reminder.completed(),
                        "completionDate": reminder.completionDate(),
                        "dueDate": reminder.dueDate(),
                        "alldayDueDate": reminder.alldayDueDate(),
                        "remindMeDate": reminder.remindMeDate(),
                        "priority": reminder.priority(),
                        "flagged": reminder.flagged()
                    }
                    
                    # Apply filters
                    if filter_params:
                        if "completed" in filter_params and reminder_dict["completed"] != filter_params["completed"]:
                            continue
                        if "flagged" in filter_params and reminder_dict["flagged"] != filter_params["flagged"]:
                            continue
                        if "priority" in filter_params and reminder_dict["priority"] != filter_params["priority"]:
                            continue
                    
                    all_reminders.append(reminder_dict)
            
            return all_reminders
        except Exception as e:
            logger.error("Failed to list reminders: %s", e)
            raise

def main():
    """Simple CLI interface"""
    service = RemindersService()
    
    if len(sys.argv) < 2:
        print("Usage: reminder.py [list|create] [args...]")
        sys.exit(1)
    
    command = sys.argv[1]
    if command == "list":
        reminders = service.list_reminders()
        print(json.dumps(reminders, indent=2))
    elif command == "create":
        if len(sys.argv) < 4:
            print("Usage: reminder.py create <name> <body>")
            sys.exit(1)
        reminder = service.create_reminder(sys.argv[2], sys.argv[3])
        print(json.dumps(reminder, indent=2))

if __name__ == "__main__":
    main() 